app/background.cc

```diff
@@ -0,0 +1,161 @@
+            }
+        } else {
+            // no more frames, but if we processed some, try to reset position and go again
+            if (pbkd->frm>0 && pbkd->cap.set(cv::CAP_PROP_POS_FRAMES, 0)) {
+                std::unique_lock<std::mutex> hold(pbkd->mux);
+                pbkd->frm = 0;
+            } else {
+                // unable to reset or previous attempt produced no more frames - stop
+                if (pbkd->dbg) fprintf(stderr, "background: thread stopping at end of stream and not resettable\n");
+                break;
+            }
+        }
+    }
+    if (pbkd->dbg) fprintf(stderr, "background: thread stop\n");
+}
+
+void *load_background(const char *path, int debug) {
+    background_t *pbkd = new background_t;
+    background_t &bkd = *pbkd;
+
+    bkd.dbg = debug;
+    bkd.cap.open(path, cv::CAP_ANY);    // explicitly ask for auto-detection of backend
+    if (!bkd.cap.isOpened()) {
+        if (bkd.dbg) fprintf(stderr, "cv::VideoCapture cannot open: %s\n", path);
+        delete pbkd;
+        return nullptr;
+    }
+    bkd.cap.set(cv::CAP_PROP_CONVERT_RGB, true);
+    int fcc = (int)bkd.cap.get(cv::CAP_PROP_FOURCC);
+    bkd.fps = bkd.cap.get(cv::CAP_PROP_FPS);
+    int cnt = (int)bkd.cap.get(cv::CAP_PROP_FRAME_COUNT);
+    // Here be the logic...
+    //  if: can read 2 video frames => it's a video
+    //  else: is loaded as an image => it's an image
+    //  else: it's not usable.
+    if (bkd.cap.read(bkd.raw) && bkd.cap.read(bkd.raw)) {
+        // it's a video, try a reset and start reader thread..
+        if (bkd.cap.set(cv::CAP_PROP_POS_FRAMES, 0))
+            bkd.frm = 0;
+        else
+            bkd.frm = 2;
+        bkd.vid = true;
+        bkd.run = true;
+        bkd.thr = std::thread(read_thread, pbkd);
+    } else {
+        // static image file, try loading..
+        bkd.cap.release();
+        bkd.raw = cv::imread(path);
+        bkd.vid = false;
+        if (bkd.raw.empty()) {
+            if (bkd.dbg) fprintf(stderr, "cv::imread cannot read: %s\n", path);
+            delete pbkd;
+            return nullptr;
+        }
+    }
+    if (bkd.dbg) fprintf(stderr, "background properties:\n\tvid: %s\n\tfcc: %08x (%.4s)\n\tfps: %f\n\tcnt: %d\n",
+        bkd.vid ? "yes":"no", fcc, (char *)&fcc, bkd.fps, cnt);
+    return pbkd;
+}
+
+int grab_background(void *handle, int width, int height, cv::Mat& out) {
+    background_t *pbkd = (background_t *)handle;
+    if (!pbkd)
+        return -1;
+    // static image or video?
+    int frm ;
+    if (pbkd->vid) {
+        // grab frame & frame no. under mutex
+        std::unique_lock<std::mutex> hold(pbkd->mux);
+        cv::resize(pbkd->raw, out, cv::Size(width, height));
+        frm = pbkd->frm;
+    } else {
+        // resize still image as requested into out
+        cv::resize(pbkd->raw, out, cv::Size(width, height));
+        frm = 1;
+    }
+    return frm;
+}
+
+void drop_background(void *handle) {
+    if (handle) {
```

Comment:
Negate condition to exit early. Reduces indentation for the proc overall.

Reply:
Thanks - yep, will fix.